tools/bootstrap_commitlogs/main/main.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/m3db/m3db/ts"
+	"github.com/m3db/m3db/x/io"
+	"github.com/m3db/m3x/instrument"
+	xlog "github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+)
+
+var (
+	pathPrefixArg         = flag.String("path-prefix", "/var/lib/m3db", "Path prefix")
+	namespaceArg          = flag.String("namespace", "metrics", "Namespace")
+	blockSizeArg          = flag.String("block-size", "2h", "Block size")
+	flushSizeArg          = flag.Int("flush-size", 524288, "Flush size of commit log")
+	bootstrapRetentionArg = flag.String("retention", "48h", "Retention")
+	shardsCountArg        = flag.Int("shards-count", 8192, "Shards count - set number too bootstrap all shards in range")
+	shardsArg             = flag.String("shards", "", "Shards - set comma separated list of shards")
+	debugListenAddressArg = flag.String("debug-listen-address", "", "Debug listen address - if set will expose pprof, i.e. ':8080'")
+)
+
+func main() {
+	flag.Parse()
+	if *pathPrefixArg == "" ||
+		*namespaceArg == "" ||
+		*blockSizeArg == "" ||
+		*bootstrapRetentionArg == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	var (
+		pathPrefix         = *pathPrefixArg
+		namespace          = *namespaceArg
+		blockSize          = *blockSizeArg
+		flushSize          = *flushSizeArg
+		bootstrapRetention = *bootstrapRetentionArg
+		shardsCount        = *shardsCountArg
+		shards             = *shardsArg
+		debugListenAddress = *debugListenAddressArg
+	)
+
+	log := xlog.NewLogger(os.Stderr)
+
+	if debugListenAddress != "" {
+		go func() {
+			err := http.ListenAndServe(debugListenAddress, http.DefaultServeMux)
+			if err != nil {
+				log.Errorf("could not start debug listen server at '%s': %v", debugListenAddress, err)
+			}
+		}()
+	}
+
+	shardTimeRanges := result.ShardTimeRanges{}
+
+	blockSizeLen, err := time.ParseDuration(blockSize)
+	if err != nil {
+		log.Fatalf("could not parse block size '%s': %v", blockSize, err)
+	}
+
+	retentionLen, err := time.ParseDuration(bootstrapRetention)
+	if err != nil {
+		log.Fatalf("could not parse retention '%s': %v", bootstrapRetention, err)
+	}
+
+	now := time.Now()
+	end := now.Truncate(blockSizeLen).Add(blockSizeLen)    // exclusive
+	start := now.Truncate(blockSizeLen).Add(-retentionLen) // inclusive
+
+	var shardsAll []uint32
+	if strings.TrimSpace(shards) != "" {
+		for _, shard := range strings.Split(shards, ",") {
+			shard = strings.TrimSpace(shard)
+			if shard == "" {
+				continue
+			}
+			value, err := strconv.Atoi(shard)
+			if err != nil {
+				log.Fatalf("could not parse shard '%s': %v", shard, err)
+			}
+			rng := xtime.Range{Start: start, End: end}
+			shardTimeRanges[uint32(value)] = xtime.NewRanges().AddRange(rng)
+			shardsAll = append(shardsAll, uint32(value))
+		}
+	} else if shardsCount > 0 {
+		for i := uint32(0); i < uint32(shardsCount); i++ {
+			rng := xtime.Range{Start: start, End: end}
+			shardTimeRanges[i] = xtime.NewRanges().AddRange(rng)
+			shardsAll = append(shardsAll, i)
+		}
+	} else {
+		flag.Usage()
```

Comment:
nit: mention explicitly the issue with bad shard configuration

Reply:
What issue?